### PR TITLE
Fail parsing dates beyond a representable range

### LINKED ISSUE #2888: Integer overflow when parsing dates beyond representable range
Correct
```
presto:default> SELECT DATE '5881580-07-11';
     _col0
----------------
 +5881580-07-11
```

wrong:
```
presto:default> SELECT DATE '5881580-07-12';
     _col0
----------------
 -5877641-06-23
```

### PR DESCRIPTION
**Pull request by Praveen2112**
Fixes #2888

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/type/DateOperators.java [hunk 1]
```diff
@@ -156,6 +157,9 @@ public static long castFromSlice(@SqlType("varchar(x)") Slice value)
         catch (IllegalArgumentException e) {
             throw new PrestoException(INVALID_CAST_ARGUMENT, "Value cannot be cast to date: " + value.toStringUtf8(), e);
         }
+        catch (ArithmeticException e) {
+            throw new PrestoException(DATE_OUT_OF_RANGE, "Out of range for Date: " + value.toStringUtf8(), e);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/type/DateOperators.java]: This should be `INVALID_CAST_ARGUMENT` as well

File: presto-spi/src/main/java/io/prestosql/spi/StandardErrorCode.java [hunk 2]
```diff
@@ -106,6 +106,7 @@
     NULL_TREATMENT_NOT_ALLOWED(83, USER_ERROR),
     INVALID_ROW_FILTER(84, USER_ERROR),
     INVALID_COLUMN_MASK(85, USER_ERROR),
+    DATE_OUT_OF_RANGE(86, USER_ERROR),            // caught by TRY
```
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/StandardErrorCode.java]: not needed, use `INVALID_CAST_ARGUMENT` instead

File: presto-main/src/test/java/io/prestosql/type/TestDateTimeOperatorsBase.java [hunk 3]
```diff
@@ -251,6 +252,13 @@ public void testIsDistinctFrom()
         assertFunction("DATE '2013-10-27' IS DISTINCT FROM NULL", BOOLEAN, true);
     }
 
+    @Test
+    public void testDateCastFromVarchar()
+    {
+        assertFunction("DATE '2013-02-02'", DATE, toDate(new DateTime(2013, 2, 2, 0, 0, 0, 0, UTC)));
+        assertInvalidFunction("DATE '5881580-07-12'", DATE_OUT_OF_RANGE, "Out of range for Date: 5881580-07-12");
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/type/TestDateTimeOperatorsBase.java]: What happens if we exceed long range in `DateTimeUtils parseDate`?

Comment (Praveen2112) [reply presto-main/src/test/java/io/prestosql/type/TestDateTimeOperatorsBase.java]: It throws INVALID_CAST_ARGUMENT

Comment (findepi) [reply presto-main/src/test/java/io/prestosql/type/TestDateTimeOperatorsBase.java]: can you add a test case for this?

File: presto-main/src/test/java/io/prestosql/type/TestDateTimeOperatorsBase.java [hunk 4]
```diff
@@ -251,6 +252,13 @@ public void testIsDistinctFrom()
         assertFunction("DATE '2013-10-27' IS DISTINCT FROM NULL", BOOLEAN, true);
     }
 
+    @Test
+    public void testDateCastFromVarchar()
+    {
+        assertFunction("DATE '2013-02-02'", DATE, toDate(new DateTime(2013, 2, 2, 0, 0, 0, 0, UTC)));
+        assertInvalidFunction("DATE '5881580-07-12'", INVALID_CAST_ARGUMENT, "Value cannot be cast to date: 5881580-07-12");
+    }
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/type/TestDateTimeOperatorsBase.java]: Can you also include a test case when value fails within `DATE_FORMATTER.parseMillis(value)` method (due to overflow)?

Comment (Praveen2112) [reply presto-main/src/test/java/io/prestosql/type/TestDateTimeOperatorsBase.java]: have applied it

**Comment by findepi** [comment]
Merged, thanks!

**Pull request merged**